### ISSUE TITLE
PostgreSQL: support UDT as a column in other UDT

### ISSUE THREAD

**Issue by remal** [jOOQ/jOOQ#12913]
### Expected behavior
`column IS DISTINCT FROM row(?)` works correctly when `column` is a UDT, where one of the columns - another UDT

### Actual behavior
I'm facing this error: `PSQLException: ERROR: could not determine data type of parameter $1`

### Steps to reproduce the problem

I have to say that I'm not eligible to provide you with the exact code, so here is some pseudo-code.

Apply these DDL statements:
```sql
CREATE TYPE inner_type AS (
  id   VARCHAR(255),
  name VARCHAR(255)
);

CREATE TYPE outer_type AS (
  inner inner_type
  -- some other columns
);

CREATE TABLE test_table (
  id SERIAL PRIMARY KEY.
  outer outer_type
);
```

After that execute jOOQ generator.

Then, execute a statement like this:
```java
var outerToExclude = new OuterTypeRecord();
dslContext.selectFrom(TEST_TABLE)
  .where(TEST_TABLE.OUTER.isDistinctFrom(outerToExclude))
  .fetch()
```

In this case, jOOQ executes a statement like this: `SELECT * FROM test_table WHERE outer IS DISTINCT FROM row(?)`. And you'll get an error I wrote above.

### Versions

- jOOQ: 3.16.2
- Java: 11.0.13
- Database (include vendor): PostgreSQL 13 (via `postgres:13-alpine` Docker container)
- OS: Windows 10
- JDBC Driver (include name if inofficial driver): from Maven Central: `org.postgresql:postgresql:42.3.1`

### Potential fix:

Add this:
```java
ctx.sql("::" + this.getDataType().getTypeName());
```

... after this line: https://github.com/jOOQ/jOOQ/blob/35a4d4936b2124a2dbb6b4191ac7d1519cdc39dd/jOOQ/src/main/java/org/jooq/impl/QualifiedRecordConstant.java#L155

This change helped in my case, but I'm sure there are a few other places in the code.

**Comment by remal** [comment]
I think so. But I haven't checked.

**Comment by lukaseder** [comment]
This is a duplicate of https://github.com/jOOQ/jOOQ/issues/13174, which was fixed in jOOQ 3.17.0, 3.16.5 (https://github.com/jOOQ/jOOQ/issues/13179), and 3.15.9 (https://github.com/jOOQ/jOOQ/issues/13180)
